connectors/sources/google_drive.py

```diff
@@ -0,0 +1,638 @@
+
+        Args:
+            retry_counter (int): Current number of attempted retries
+            log_message (str): Warning message logged
+
+        Returns:
+            int: increased retry_counter
+        """
+        retry_counter += 1
+        if retry_counter > self.retry_count:
+            raise
+        logger.warning(log_message)
+        await asyncio.sleep(DEFAULT_WAIT_MULTIPLIER**retry_counter)
+        return retry_counter
+
+
+class GoogleDriveDataSource(BaseDataSource):
+    """Google Drive"""
+
+    name = "Google Drive"
+    service_type = "google_drive"
+
+    def __init__(self, configuration):
+        """Set up the connection to the Google Drive Client.
+
+        Args:
+            configuration (DataSourceConfiguration): Object of DataSourceConfiguration class.
+        """
+        super().__init__(configuration=configuration)
+
+    @classmethod
+    def get_default_configuration(cls):
+        """Get the default configuration for Google Drive.
+
+        Returns:
+            dict: Default configuration.
+        """
+        default_credentials = {
+            "type": "service_account",
+            "project_id": "project_id",
+            "private_key_id": "abc",
+            "private_key": open(
+                os.path.abspath(DEFAULT_PEM_FILE)
+            ).read(),  # TODO: change this and provide meaningful defaults
+            "client_email": "123-abc@developer.gserviceaccount.com",
+            "client_id": "123-abc.apps.googleusercontent.com",
+            "auth_uri": "https://accounts.google.com/o/oauth2/auth",
+            "token_uri": "http://localhost:10339/token",
+        }
+        return {
+            "service_account_credentials": {
+                "display": "textarea",
+                "label": "Google Drive service account JSON",
+                "order": 1,
+                "type": "str",
+                "value": json.dumps(default_credentials),
+            },
+            "retry_count": {
+                "default_value": DEFAULT_RETRY_COUNT,
+                "display": "numeric",
+                "label": "Maximum retries for failed requests",
+                "order": 2,
+                "required": False,
+                "type": "int",
+                "ui_restrictions": ["advanced"],
+                "value": DEFAULT_RETRY_COUNT,
+            },
+        }
+
+    @cached_property
+    def _google_drive_client(self):
+        """Initialize and return the GoogleDriveClient
+
+        Returns:
+            GoogleDriveClient: An instance of the GoogleDriveClient.
+        """
+        json_credentials = json.loads(self.configuration["service_account_credentials"])
+
+        if (
+            json_credentials.get("private_key")
+            and "\n" not in json_credentials["private_key"]
+        ):
+            json_credentials["private_key"] = get_pem_format(
+                key=json_credentials["private_key"].strip(),
+                max_split=2,
```

Comment:
Can you also validate the private_key? Will `get_pem_format` raise an error if private key is invalid?

Reply:
`get_pem_format` is actually not needed - I copy pasted this from google cloud storage client. It was used to handle default credentials.

I realised we can use `connecotrs.json` for providing credentials for ftests. 

`get_pem_format` just formats the private key into single line format. I would expect that `aiogoogle` or `google api` will error out if private key is invalid. User needs to only copy-paste the generated json here.